### PR TITLE
Refactor Unification Loop

### LINKED ISSUE #2396: Optimize unification inner loop
The inner loop of the unification code has poor performance because using `MaybeT` involves sequencing many binds (`>>=`) in the underlying monad. We will separate the code which identifies unification problems from the code which solves the unification problem. A simplified schematic of the current code is:

``` hs
unify
    :: TermLike variable
    -> TermLike variable
    -> MaybeT unifier (Pattern variable)
unify term1 term2 = 
    matchAndUnify1 term1 term2
    <|> matchAndUnify2 term1 term2
    <|> matchAndUnify3 term1 term2
    ...
```

`unify` could mean `maybeTermAnd` or `maybeTermEquals`.

We will split each `matchAndUnifyN` into a pair:

``` hs
matchN :: TermLike variable -> TermLike variable -> Maybe problem
matchN = ...
{-# INLINE matchN #-}

unifyN :: problem -> unifier (Pattern variable)
```

`problem` stands for a data type representing each type of unification problem. It will be different for each pair of new functions. It is important that the `matchN` functions be inlined to make the pattern matching code more efficient, but they should be very small functions anyway. `unify` will take this simplified form:

``` hs
unify term1 term2
  | Just problem1 <- match1 term1 term2 = unify1 problem1 & lift
  | Just problem2 <- match2 term1 term2 = unify2 problem2 & lift
  | Just problem3 <- match3 term1 term2 = unify3 problem3 & lift
  ...
  | otherwise = empty
```

### PR DESCRIPTION
**Pull request by emarzion**
Fixes #2396

Replacing branch #2535 

The code for [maybeTermEquals](https://github.com/kframework/kore/blob/master/kore/src/Kore/Step/Simplification/AndTerms.hs#L130) and [maybeTermAnd](https://github.com/kframework/kore/blob/master/kore/src/Kore/Step/Simplification/AndTerms.hs#L130) are being modified as follows:  Each `asum` is replaced by a list of patterns + function calls, with each invoked function being replaced by two functions, the first corresponding to the pattern match, which returns the necessary data for the second function which performs the unification, now accepting the data from the match as a single argument.   The current conventions for returning the data in this PR are:

1. If there are two or more possible matches corresponding to one of these functions, the data is returned as part of a custom sum type with two or more constructors; otherwise, it is returned as is.
2. If, within a single match, only one piece of data is returned, then it is returned as is (without being wrapped; the exception to this is if the type is too verbose in which case a newtype is used); otherwise, the data is packaged and returned together in a custom record type.
3. If a function has only a single match and returns no data, then `()` is returned (since we are returning `Maybe ()` it may be better to just return a `Bool`, but in the interest of regularity I've chosen the former option).

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

File: kore/src/Kore/Builtin/Bool.hs [hunk 1]
```diff
@@ -163,48 +165,63 @@ builtinFunctions =
...
+    }
+
+matchBools ::
+    TermLike RewritingVariableName ->
+    TermLike RewritingVariableName ->
+    Maybe UnifyBool
+matchBools first second
+    | InternalBool_ bool1 <- first
+      , InternalBool_ bool2 <- second =
+        Just $ UnifyBool bool1 bool2
```
Comment (ttuegel) [inline kore/src/Kore/Builtin/Bool.hs:179]: If we are going to use name shadowing, let's do it properly with named field puns:

```suggestion
        Just $ UnifyBool{bool1, bool2}
```

File: kore/src/Kore/Builtin/Bool.hs [hunk 2]
```diff
@@ -163,48 +165,63 @@ builtinFunctions =
...
+    UnifyBool ->
+    unifier (Pattern RewritingVariableName)
+unifyBool termLike1 termLike2 unifyData =
+    if bool1 == bool2
+        then return (Pattern.fromTermLike termLike1)
+        else
+            Unify.explainAndReturnBottom
+                "different Bool domain values"
+                termLike1
+                termLike2
```
Comment (ttuegel) [inline kore/src/Kore/Builtin/Bool.hs:198]:
```suggestion
unifyBool termLike1 termLike2 unifyData
  | bool1 == bool2 = return (Pattern.fromTermLike termLike1)
  | otherwise =
    Unify.explainAndReturnBottom
        "different Bool domain values"
        termLike1
        termLike2
```

File: kore/src/Kore/Builtin/Bool.hs [hunk 3]
```diff
@@ -163,48 +165,63 @@ builtinFunctions =
...
-                            termLike1
-                            termLike2
-    worker _ _ = empty
+    UnifyBool{bool1, bool2} = unifyData
+
+matchUnifyBoolAnd ::
+    TermLike RewritingVariableName ->
+    TermLike RewritingVariableName ->
+    Maybe (BoolAnd (TermLike RewritingVariableName))
+matchUnifyBoolAnd first second
```
Comment (ttuegel) [inline kore/src/Kore/Builtin/Bool.hs:206]: Where is the `UnifyBoolAnd` type?

File: kore/src/Kore/Builtin/Bool.hs [hunk 4]
```diff
@@ -231,44 +248,60 @@ unifyBothWith unify termLike1 operand1 operand2 = lift $ do
     unify' term1 term2 =
         Pattern.withoutTerm <$> unify term1 term2
 
+matchUnifyBoolOr ::
+    TermLike RewritingVariableName ->
+    TermLike RewritingVariableName ->
+    Maybe (BoolOr (TermLike RewritingVariableName))
```
Comment (ttuegel) [inline kore/src/Kore/Builtin/Bool.hs:254]: Do `BoolAnd` and `BoolOr` really need this type parameter? Isn't it always `TermLike RewritingVariableName`?

File: kore/src/Kore/Builtin/Int.hs [hunk 5]
```diff
@@ -424,24 +428,57 @@ matchIntEqual =
...
-    TermLike variable ->
-    MaybeT unifier (Pattern variable)
-unifyInt term1@(InternalInt_ int1) term2@(InternalInt_ int2) =
-    assert (on (==) internalIntSort int1 int2) $ lift worker
+    TermLike RewritingVariableName ->
+    TermLike RewritingVariableName ->
+    UnifyInt ->
+    unifier (Pattern RewritingVariableName)
+unifyInt term1 term2 unifyData =
+    assert (on (==) internalIntSort int1 int2) worker -- should this be part of match?
```
Comment (ttuegel) [inline kore/src/Kore/Builtin/Int.hs:457]: Yes, this assertion should be part of the matching function.

File: kore/src/Kore/Step/Simplification/ExpandAlias.hs [hunk 6]
```diff
@@ -34,25 +27,21 @@ import Kore.Internal.TermLike (
...
+    { term1, term2 :: !(TermLike RewritingVariableName)
+    }
+
+matchExpandAlias ::
     TermLike RewritingVariableName ->
     TermLike RewritingVariableName ->
-    MaybeT unifier (Pattern RewritingVariableName)
-expandAlias recurse t1 t2 =
+    Maybe UnifyExpandAlias
+matchExpandAlias t1 t2 =
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/ExpandAlias.hs:40]: What is the corresponding function for this one, that carries out unification?

File: kore/src/Kore/Step/Simplification/InjSimplifier.hs [hunk 7]
```diff
@@ -66,11 +80,12 @@ data InjSimplifier = InjSimplifier
       --
       --        Returns 'Distinct' if the sort injections cannot match, or 'Unknown' if
       --        further simplification could produce matching injections.
-      unifyInj ::
+      unifyInjs ::
         forall variable.
         InternalVariable variable =>
         Inj (TermLike variable) ->
         Inj (TermLike variable) ->
+        Either Distinct InjUnify ->
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/InjSimplifier.hs:88]: `InjUnify` should contain all the data which `unifyInjs` needs to do its job. This will minimize the risk that `unifyInjs` is called with the wrong `Inj`s for the supplied match data.

File: kore/src/Kore/Step/Simplification/NoConfusion.hs [hunk 8]
```diff
@@ -27,6 +25,31 @@ import Prelude.Kore hiding (
     concat,
  )
 
+data UnifyEqualInjectiveHeadsAndEquals = UnifyEqualInjectiveHeadsAndEquals
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/NoConfusion.hs:28]:
```suggestion
data UnifyEqualInjectiveHeads = UnifyEqualInjectiveHeads
```

File: kore/src/Kore/Step/Simplification/NoConfusion.hs [hunk 9]
```diff
@@ -27,6 +25,31 @@ import Prelude.Kore hiding (
...
+matchEqualInjectiveHeadsAndEquals ::
+    TermLike RewritingVariableName ->
+    TermLike RewritingVariableName ->
+    Maybe UnifyEqualInjectiveHeadsAndEquals
+matchEqualInjectiveHeadsAndEquals first second
+    | App_ firstHead firstChildren <- first
+      , App_ secondHead secondChildren <- second
+      , Symbol.isInjective firstHead
+      , Symbol.isInjective secondHead
+      , firstHead == secondHead --is one of the above redundant in light of this?
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/NoConfusion.hs:43]: You can remove `Symbol.isInjective secondHead` above, and leave a comment explaining.

File: kore/src/Kore/Step/Simplification/NoConfusion.hs [hunk 10]
```diff
@@ -39,29 +62,50 @@ equalInjectiveHeadsAndEquals ::
...
+        UnifyEqualInjectiveHeadsAndEquals
+            { firstHead
+            , firstChildren
+            , secondChildren
+            } = unifyData
+
+-- data ConstructorAndEqualsAssumesDifferentHeads = ConstructorAndEqualsAssumesDifferentHeads
+--     { firstHead, secondHead :: Symbol
+--     }
+
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/NoConfusion.hs:91]:
```suggestion
```

File: kore/src/Kore/Step/Simplification/NoConfusion.hs [hunk 11]
```diff
@@ -39,29 +62,50 @@ equalInjectiveHeadsAndEquals ::
...
+matchConstructorAndEqualsAssumesDifferentHeads ::
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/NoConfusion.hs:92]: This function no longer assumes anything.

```suggestion
matchDifferentConstructors ::
```

File: kore/src/Kore/Step/Simplification/Overloading.hs [hunk 12]
```diff
@@ -127,6 +132,158 @@ matchOverloading termPair =
             Simple pair -> return pair
             _ -> notApplicable
 
+-- flipPairBack :: Pair a -> Pair a
+-- flipPairBack (Pair x y) = Pair y x
+
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/Overloading.hs:137]:
```suggestion
```

File: kore/src/Kore/Step/Simplification/Overloading.hs [hunk 13]
```diff
@@ -1,16 +1,21 @@
+{-# LANGUAGE PartialTypeSignatures #-}
+
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/Overloading.hs:2]:
```suggestion
```

File: kore/src/Kore/Step/Simplification/Overloading.hs [hunk 14]
```diff
@@ -127,6 +132,158 @@ matchOverloading termPair =
...
+    , injChild :: TermLike RewritingVariableName
+    }
+
+data OverloadedConstructorSortInjectionAndEquals
+    = Overload1 UnifyOverload1
+    | Overload2 UnifyOverload2
+    | Overload3 UnifyOverload3
+    | Overload4 UnifyOverload4
+    | Overload5 UnifyOverload5
+    | Overload6 UnifyOverload6
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/Overloading.hs:180]: These all need real names.

**Review by ttuegel — CHANGES_REQUESTED** [review]
In addition, all of the new top-level definitions need documentation. For the unification cases (which is most of it) this could be as simple as showing the Kore terms that are being unified.

File: kore/src/Kore/Builtin/Bool.hs (hunk 3 above)
Comment (emarzion) [reply kore/src/Kore/Builtin/Bool.hs:206]: There is none, here I am following the convention that if the match is returning only a single datum that we just return that without wrapping.

File: kore/src/Kore/Builtin/Bool.hs (hunk 4 above)
Comment (emarzion) [reply kore/src/Kore/Builtin/Bool.hs:254]: From the looks of it, they probably don't anymore, I'm guessing this might be something left over from the `variable` to `RewritingVariableName` change?  In any case I'll go ahead and change those.

File: kore/src/Kore/Step/Simplification/ExpandAlias.hs (hunk 6 above)
Comment (emarzion) [reply kore/src/Kore/Step/Simplification/ExpandAlias.hs:40]: The terms are fed back into `maybeTermAnd` [here](https://github.com/kframework/kore/blob/2a9962c6b9b742952e4ccdb134ae42b823ff8b4f/kore/src/Kore/Step/Simplification/AndTerms.hs#L238).

File: kore/src/Kore/Step/Simplification/Overloading.hs (hunk 14 above)
Comment (emarzion) [reply kore/src/Kore/Step/Simplification/Overloading.hs:180]: This code isn't currently being used in this PR, should I just remove it for now?

File: kore/src/Kore/Builtin/Bool.hs [hunk 15]
```diff
@@ -163,48 +165,69 @@ builtinFunctions =
...
-    InternalVariable variable =>
+    forall unifier.
     MonadUnify unifier =>
-    TermLike variable ->
-    TermLike variable ->
-    MaybeT unifier (Pattern variable)
-unifyBool a b =
-    worker a b <|> worker b a
+    TermLike RewritingVariableName ->
+    TermLike RewritingVariableName ->
```
Comment (ttuegel) [inline kore/src/Kore/Builtin/Bool.hs:189]: We should not pass in the original `TermLike`s to any `unify...` function.

File: kore/src/Kore/Builtin/Int.hs [hunk 16]
```diff
@@ -424,24 +428,58 @@ matchIntEqual =
...
+
+-- | Matches two Int values that have equals sorts.
+matchInt ::
+    TermLike RewritingVariableName ->
+    TermLike RewritingVariableName ->
+    Maybe UnifyInt
+matchInt first second
+    | InternalInt_ int1 <- first
+      , InternalInt_ int2 <- second
+      , on (==) internalIntSort int1 int2 =
```
Comment (ttuegel) [inline kore/src/Kore/Builtin/Int.hs:443]: I think this should be an assertion. Attempting to unify two things with different sorts doesn't make any sense.

File: kore/src/Kore/Builtin/Int.hs [hunk 17]
```diff
@@ -424,24 +428,58 @@ matchIntEqual =
...
+    UnifyInt{int1, int2} = unifyData
+
+data UnifyIntEq = UnifyIntEq
+    { eqTerm :: !(EqTerm (TermLike RewritingVariableName))
+    , value :: !Bool
+    }
+
+{- | Matches two terms when second is a bool term
+    and the first is a function pattern matching
+    the @INT.eq@ hooked symbol.
```
Comment (ttuegel) [inline kore/src/Kore/Builtin/Int.hs:470]: I think we can make all of this documentation clearer using the Kore notation:

```suggestion
{- | Matches
@
\\equals{_, _}(\\dv{Bool}(_), eqInt{_}(_, _))
@
```

File: kore/src/Kore/Step/Simplification/AndTerms.hs [hunk 18]
```diff
@@ -112,10 +113,12 @@ termUnification notSimplifier = \term1 term2 ->
         TermLike RewritingVariableName ->
         unifier (Pattern RewritingVariableName)
     termUnificationWorker pat1 pat2 = do
+        injSimplifier <- Simplifier.askInjSimplifier
+        OverloadSimplifier{isOverloaded} <- Simplifier.askOverloadSimplifier
         let maybeTermUnification ::
                 MaybeT unifier (Pattern RewritingVariableName)
             maybeTermUnification =
-                maybeTermAnd notSimplifier termUnificationWorker pat1 pat2
+                maybeTermAnd notSimplifier termUnificationWorker injSimplifier isOverloaded pat1 pat2
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/AndTerms.hs:121]: Why are we passing the new parameters? `maybeTermAnd` can retrieve these from its `unifier` monad.

File: kore/src/Kore/Step/Simplification/AndTerms.hs [hunk 19]
```diff
@@ -647,14 +826,21 @@ compareForEquals first second
...
+      , bytesFirst /= bytesSecond =
+        Just ()
+    | otherwise = Nothing
+{-# INLINE matchBytesDifferent #-}
+
+bytesDifferent ::
+    MonadUnify unifier =>
+    unifier (Pattern RewritingVariableName)
+bytesDifferent =
+    return Pattern.bottom
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/AndTerms.hs:846]: This code would be simplified if we did not assume that the `Bytes` are different, but checked.

File: kore/src/Kore/Step/Simplification/Overloading.hs [hunk 20]
```diff
@@ -465,7 +465,7 @@ mkInj ::
     Inj () ->
     TermLike RewritingVariableName ->
     TermLike RewritingVariableName
-mkInj inj injChild = (synthesize . InjF) inj{injChild}
+mkInj inj injChild = (synthesize . InjF) (inj :: Inj ()){injChild}
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/Overloading.hs:468]: Why is this necessary? The type of `inj` should be fixed by the function's signature.

**Review by ttuegel — CHANGES_REQUESTED** [review]

File: kore/src/Kore/Step/Simplification/AndTerms.hs (hunk 18 above)
Comment (emarzion) [reply kore/src/Kore/Step/Simplification/AndTerms.hs:121]: They are used in some match functions (`matchSortInjectionAndEquals` for instance) so we do not have access at that point to the `unifier` monad.

File: kore/src/Kore/Step/Simplification/AndTerms.hs (hunk 19 above)
Comment (emarzion) [reply kore/src/Kore/Step/Simplification/AndTerms.hs:846]: I don't understand, is this not being done on line 837?

File: kore/src/Kore/Builtin/Bool.hs (hunk 15 above)
Comment (emarzion) [reply kore/src/Kore/Builtin/Bool.hs:200]: We are putting this off for the time being per our discussion at the standup about the `explainAndReturnBottom` function.

File: kore/src/Kore/Step/Simplification/AndTerms.hs [hunk 21]
```diff
@@ -647,14 +920,38 @@ compareForEquals first second
...
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/AndTerms.hs:957]:
```suggestion
matchBytesDifferent first second
    | _ :< InternalBytesF (Const bytesFirst) <- Recursive.project first
      , _ :< InternalBytesF (Const bytesSecond) <- Recursive.project second =
        Just ()
    | otherwise = Nothing
{-# INLINE matchBytesDifferent #-}

bytesDifferent ::
    MonadUnify unifier =>
    unifier (Pattern RewritingVariableName)
bytesDifferent UnifyBytes{ bytesFirst, bytesSecond }
    | bytesFirst == bytesSecond =
        return $ Pattern.fromTermLike $ internalBytes bytesFirst
    | otherwise =
        empty
```

File: kore/src/Kore/Step/Simplification/AndTerms.hs [hunk 22]
```diff
@@ -647,14 +920,40 @@ compareForEquals first second
...
+    | otherwise = Nothing
+{-# INLINE matchBytes #-}
+
+bytesDifferent ::
+    MonadUnify unifier =>
+    UnifyBytes ->
+    unifier (Pattern RewritingVariableName)
+bytesDifferent UnifyBytes{ bytes1, bytes2 }
+    | bytes1 == bytes2 =
+        return $ Pattern.fromTermLike $ mkInternalBytes' bytes1
```
Comment (emarzion) [inline kore/src/Kore/Step/Simplification/AndTerms.hs:957]: Is it better to use a `mk` function instead of returning the original term?

**Review by ttuegel — APPROVED** [review]
